### ISSUE TITLE
Unsure cell probabilities resulting in 0 MT genes

### ISSUE THREAD

**Issue by mariecrane** [broadinstitute/CellBender#178]
I have a somewhat problematic dataset that CellRanger had a hard time calling cells on (it predicted 69,035 cells when I was expecting 4000-7000). Below is the barcode rank plot:

![MC6_barcode_rank](https://user-images.githubusercontent.com/35476290/220157913-c8f75ab3-3795-4c24-84f6-149edd015c28.png)

As you can see there is not a clear knee, but I estimated 2000 expected cells and 40,000 total droplets for CellBender. The training plot was very wobbly so I decreased the learning rate until it looked better. My final input was: `cellbender remove-background --input MGI3842_GLAE-230117_MC_10X_pool1-MC6/outs/raw_feature_bc_matrix.h5 --output 00_CellBender/MC6_3/CB_filtered_feature_bc_matrix.h5 --cuda --expected-cells 2000 --total-droplets-included 40000 --fpr 0.01 --epochs 150 --learning-rate 0.000025`, which produced the following pdf output: 

[MC6_CB_filtered_feature_bc_matrix.pdf](https://github.com/broadinstitute/CellBender/files/10786236/MC6_CB_filtered_feature_bc_matrix.pdf)

While the learning plot looks okay, the cell probabilities are all over the place. In your documentation, it states "A converged inference procedure should result in the vast majority of cell probabilities being very close to either zero or one", but there's no guidance on what to do when there are lots of droplets with uncertain cell probabilities. 

The CellBender-filtered dataset did have a much more reasonable number of cells (11,905) but when I looked at the percent of mitochondrial reads, I was surprised to find that all of the cells had 0 mitochondrial reads. I know mitochondrial reads would count as background and that CellBender often reduces the percent of mitochondrial reads, but this seems a bit extreme. What steps can I take to improve the cell-calling probabilities and address the strange absence of mt genes?

**Comment by sjfleming** [comment]
Hi @mariecrane , good questions.  

What happens if you try this?  Still a wobbly learning curve?
```console
cellbender remove-background \
    --input MGI3842_GLAE-230117_MC_10X_pool1-MC6/outs/raw_feature_bc_matrix.h5 \
    --output 00_CellBender/MC6_3/CB_filtered_feature_bc_matrix.h5 \
    --cuda \
    --expected-cells 2000 \
    --total-droplets-included 25000 \
    --fpr 0.01 \
    --low-count-threshold 500
```
I am wondering if the `--low-count-threshold` might help, since the amount of ambient RNA is quite high.

There are a couple possibilities when it comes to the cell calling and the MT counts.
1. Maybe the cell calling could improve with the `--low-count-threshold 500` and the `--total-droplets-included 25000`, since this tells CellBender that all droplets past 25000 are empty.
2. I do think this is a hard dataset.  The PCA plot of the latent representation of gene expression shows that it's very nearly one large blob.  Now, if that is due to over-calling of cells, and if the suggestion above improves cell calling, then you might see some structure emerge in the PCA plot.  Or if this experiment only contains one cell type, then this might make sense.  But the PCA plot might be hinting at something else: it's possible this experiment just did not go well.  CellBender uses UMI counts and gene expression together to make cell calls.  And it seems like the expression profiles (based on guessing from the PCA plot) are not very diverse, and are similar to empty droplets.
3. As for the MT counts, it does seem like complete disappearance is probably an overcorrection.  (After all, the MT counts had to come from somewhere.)  The upcoming version 0.3.0 release will be a little more conservative, and might help with this kind of overcorrection.  However, again, this is a hard dataset.  The number of UMI counts in empty droplets is around 1000.  Most cells seem to have between 2000 and 3000 UMI counts.  This is very very challenging from the standpoint of noise removal.  While I wouldn't expect to see zero MT counts after noise removal, I would probably expect to see MT counts very drastically reduced.

**Comment by mariecrane** [comment]
Thanks @sjfleming for your thorough response! I tried rerunning it with the parameters you suggested and the learning curve was still wobbly (see attached PDF). 

[MC6_followup_CB_filtered_feature_bc_matrix.pdf](https://github.com/broadinstitute/CellBender/files/10806941/MC6_followup_CB_filtered_feature_bc_matrix.pdf)

Would increasing `--learning-rate` be the next step here? Or should I continue to increase `--low-count-threshold`? I'm still not entirely sure how to determine the value of `--low-count-threshold` to use, or when it's appropriate to deviate from the default. 

Thank you for your help!

**Comment by mariecrane** [comment]
Oops, I didn't mean to close it (I haven't submitted an issue to a github repo before), sorry about that.

**Comment by sjfleming** [comment]
Oh wow, that did not fix things at all...

You could try `--low-count-threshold 900` perhaps.  If you decrease the learning rate, then the learning curve may get smoother.  But I actually don't think the wobbly learning curve is the main problem here.  I think that cellbender is having a really hard time calling cells.  I'm not sure why, but I think it's just a difficult dataset in that the cells have 2000-3000 UMI counts, and the empties have 1000 counts.

Typically you do not have to tweak `--low-count-threshold` unless the ambient RNA is so high that cellbender's heuristics for identifying the average number of UMI counts in empty droplets.  If you look in the log file, I would be curious to know how many counts cellbender thinks empty droplets have (within the first 20 or so lines of the log file).  If that number is way off, then changing `--low-count-threshold` is one way to try to get cellbender back on track.  If that number is reasonable (around 1000 or more), then cellbender is doing a fine job of identifying that prior on empty counts.  In that case, tweaking `--low-count-threshold` will not help.

If you find that the prior on empty droplet counts seems correct (and so tweaking `--low-count-threshold` is not likely to help anything), then I am not totally sure how to proceed.

One last-ditch thing you could try would be to reduce `--total-droplets-included` to the maximum number of cells you think there are in this dataset.  Here I might use 8000 or 9000 or 10000.  This will tell cellbender that, for sure, everything past `--total-droplets-included` is an empty droplet.  This will force cellbender to make use of that information when fitting its model and removing noise.  You can try to compare the output (denoised count matrix) with the output you get when you don't do this.  But I expect the output to be pretty similar.

**Comment by mariecrane** [comment]
Ah yes, the prior on empty droplets was 1095 with my original parameters and it didn't change with `--low-count-threshold 500`. So it seems like that isn't the issue here. Good to know how to interpret that parameter though! 

I'll keep tweaking the `total-droplets-included` value but it may be that this sample is simply too noisy to use. I'll go ahead and close this issue since there's not much more we can do. Thanks again for your help!

**Comment by sjfleming** [comment]
Sure thing!  Sorry it didn't help.  I agree that this dataset might be pushing past the limits of what cellbender can do a good job on.  (I guess it's producing an output, but it's probably overcorrecting, and all the MT genes are getting removed.)
